### PR TITLE
Fix searchpath handling for tsql udfs inside fmgr

### PR DESCRIPTION
**Pull request by tanscorpio7**
### Description

Currently we override the search_path for pltsql UDFs directly using the GUC variable & avoid set_config.
This leads to crash if there was another set_config('search_path... when we were in this overridden state.
Reason for crash is that the new set_config call may push the current value (the one we set in fmgr) onto
GUC stack. Now it pssoble that we exit the function call and released fcache, but fcache->prosearchpath
is still referenced in GUC stack.

To fix this we fallback to using set_config for setting the search_path & skip the check hook for search path
when setting from inside fmgr which helps with performance.
Also increase the GUC nest level since GUC_ACTION_SAVE is not possible on top of GUC_ACTION_SET
(user set), which is possible in this case. New nest level also means that we do not have to reset the search
path at the end of execution and let AtEOXact_GUC(commit=true) handle it.

We also move the set search path block in fmgr inside TRY, to ensure we reset the skip check search path
variable in case of errors.

### Issues Resolved

[BABEL-4877]

### Sign Off

Signed-off-by: Tanzeel Khan <tzlkhan@amazon.com>

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/utils/fmgr/fmgr.c [hunk 1]
```diff
@@ -803,9 +804,12 @@ fmgr_security_definer(PG_FUNCTION_ARGS)
 
 	if (fcache->prosearchpath)
 	{
-		old_search_path = namespace_search_path;
-		namespace_search_path = fcache->prosearchpath;
-		assign_search_path(fcache->prosearchpath, newextra);
+		pltsql_save_nestlevel = NewGUCNestLevel();
+		pltsql_check_search_path = false;
+		(void) set_config_option("search_path", fcache->prosearchpath,
```
Comment (surendravishnoi) [inline src/backend/utils/fmgr/fmgr.c]: see if we can move it to try/catch

File: src/backend/catalog/namespace.c [hunk 2]
```diff
@@ -4450,6 +4450,13 @@ check_search_path(char **newval, void **extra, GucSource source)
 	char	   *rawname;
 	List	   *namelist;
 
+	/* quick exit for babelfish when setting search path in fmgr_security_definer */
+	if (sql_dialect == SQL_DIALECT_TSQL && set_local_schema_for_func_hook
```
Comment (surendravishnoi) [inline src/backend/catalog/namespace.c:4454]: validate if dialect check is correct here in case of PG to TSQL call

**Review by surendravishnoi — COMMENTED** [review]
Please add problem description and solution in commit message

File: src/backend/catalog/namespace.c (hunk 2 above)
Comment (tanscorpio7) [reply src/backend/catalog/namespace.c:4454]: we set the dialect in fmgr before calling set_cofig searchpath.
and we only change search path for tsql user defined functions. so dialect will be tsql.

File: src/backend/utils/fmgr/fmgr.c (hunk 1 above)
Comment (tanscorpio7) [reply src/backend/utils/fmgr/fmgr.c]: moved inside try block

**Review by surendravishnoi — APPROVED** [review]

**Pull request merged**